java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java

```diff
@@ -0,0 +1,141 @@
+    .typeDefinition(TypeCriteria.subtypeOf("java.util.Map"))
+    .name("keySet");
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.FOR_EACH_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (hasSemantic()) {
+      ForEachStatement forEachTree = (ForEachStatement) tree;
+      ExpressionTree expressionTree = forEachTree.expression();
+      if (expressionTree.is(Tree.Kind.METHOD_INVOCATION)) {
+        MethodInvocationTree methodTree = (MethodInvocationTree) expressionTree;
+        Symbol mapSymbol = getMapSymbol(methodTree);
+        if (mapSymbol != null && MAP_KEYSET_METHOD.matches(methodTree)) {
+          new GetUsageVisitor().isCallingGetWithSymbol(forEachTree, forEachTree.variable().symbol(), mapSymbol);
+        }
+      }
+    }
+  }
+
+  @CheckForNull
+  private Symbol getMapSymbol(MethodInvocationTree tree) {
+    ExpressionTree expressionTree = tree.methodSelect();
+    // direct invocation: symbol is implicitly this
+    if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return tree.symbol().owner();
+    }
+    expressionTree = ((MemberSelectExpressionTree) expressionTree).expression();
+    if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return ((IdentifierTree) expressionTree).symbol();
+    } else {
+      return isFieldAccessedUsingSuperOrThis(expressionTree);
+    }
+  }
+
+  @CheckForNull
+  private Symbol isFieldAccessedUsingSuperOrThis(ExpressionTree expressionTree) {
+    if (expressionTree.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree memberSelectTree = (MemberSelectExpressionTree) expressionTree;
+      if (memberSelectTree.expression().is(Tree.Kind.IDENTIFIER)) {
+        IdentifierTree identifierTree = (IdentifierTree) memberSelectTree.expression();
+        if ("super".equals(identifierTree.identifierToken().text()) || "this".equals(identifierTree.identifierToken().text())) {
+          return memberSelectTree.identifier().symbol();
+        }
+      }
+    }
+    return null;
+  }
+
+  private class GetUsageVisitor extends BaseTreeVisitor {
+    private Symbol variable;
+    private boolean result;
+    private Symbol mapSymbol;
+
+    public void isCallingGetWithSymbol(ForEachStatement forEachTree, Symbol variable, Symbol mapSymbol) {
+      this.variable = variable;
+      result = false;
+      this.mapSymbol = mapSymbol;
+      scan(forEachTree.statement());
+      if (result) {
+        addIssue(forEachTree, "Iterate over the \"entrySet\" instead of the \"keySet\".");
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (MAP_GET_METHOD.matches(tree)) {
+        Tree firstArgument = Iterables.getOnlyElement(tree.arguments());
+        if (mapSymbol.equals(getMapSymbol(tree)) && firstArgument.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) firstArgument).symbol().equals(variable)) {
+          result = true;
```

Comment:
You can probably nterrupt the exploration at this point with an empty `return;`.